js/src/ipc/writer/binary.ts

```diff
@@ -0,0 +1,705 @@
+    public visitBool           (vector: Vector<Bool>)            { return this.visitBoolVector(vector);      }
+    public visitInt            (vector: Vector<Int>)             { return this.visitFlatVector(vector);      }
+    public visitFloat          (vector: Vector<Float>)           { return this.visitFlatVector(vector);      }
+    public visitUtf8           (vector: Vector<Utf8>)            { return this.visitFlatListVector(vector);  }
+    public visitBinary         (vector: Vector<Binary>)          { return this.visitFlatListVector(vector);  }
+    public visitDate           (vector: Vector<Date_>)           { return this.visitFlatVector(vector);      }
+    public visitTimestamp      (vector: Vector<Timestamp>)       { return this.visitFlatVector(vector);      }
+    public visitTime           (vector: Vector<Time>)            { return this.visitFlatVector(vector);      }
+    public visitDecimal        (vector: Vector<Decimal>)         { return this.visitFlatVector(vector);      }
+    public visitInterval       (vector: Vector<Interval>)        { return this.visitFlatVector(vector);      }
+    public visitList           (vector: Vector<List>)            { return this.visitListVector(vector);      }
+    public visitStruct         (vector: Vector<Struct>)          { return this.visitNestedVector(vector);    }
+    public visitFixedSizeBinary(vector: Vector<FixedSizeBinary>) { return this.visitFlatVector(vector);      }
+    public visitFixedSizeList  (vector: Vector<FixedSizeList>)   { return this.visitListVector(vector);      }
+    public visitMap            (vector: Vector<Map_>)            { return this.visitNestedVector(vector);    }
+    public visitDictionary     (vector: DictionaryVector)        {
+        // Dictionary written out separately. Slice offset contained in the indices
+        return this.visit(vector.indices);
+    }
+    public visitUnion(vector: Vector<DenseUnion | SparseUnion>) {
+        const { data, type, length } = vector;
+        const { offset: sliceOffset, typeIds } = data;
+        // All Union Vectors have a typeIds buffer
+        this.addBuffer(typeIds);
+        // If this is a Sparse Union, treat it like all other Nested types
+        if (type.mode === UnionMode.Sparse) {
+            return this.visitNestedVector(vector);
+        } else if (type.mode === UnionMode.Dense) {
+            // If this is a Dense Union, add the valueOffsets buffer and potentially slice the children
+            const valueOffsets = (data as DenseUnionData).valueOffsets;
+            if (sliceOffset <= 0) {
+                // If the Vector hasn't been sliced, write the existing valueOffsets
+                this.addBuffer(valueOffsets);
+                // We can treat this like all other Nested types
+                return this.visitNestedVector(vector);
+            } else {
+                // A sliced Dense Union is an unpleasant case. Because the offsets are different for
```

Comment:
Indeed... I recall that writing the unit tests to exercise this code path was a pain. Is this being tested now?

Reply:
Heh, I lifted most of the Union serialization from the C++ implementation ([comments and all](https://github.com/apache/arrow/blob/fcc13f5088323989d925270e7850b817bde19d48/cpp/src/arrow/ipc/writer.cc#L417)). I haven't added tests for them yet, but meant to generate some test data from https://github.com/apache/arrow/pull/987. I'll try that here shortly.